Display message on course outline in Studio when course contains ORA 1

muhammad-ammar opened the pull request:
TNL-2303

Sandbox: http://studio.tnl2303.m.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course

muhammad-ammar commented:
@marcotuts This is work in progress. I need your feedback on how the UI should look like:
* what should be the style of ora1 components list
* where should we place the link to `Studio Advanced Settings page`

@explorerleslie Anything you want to add from UI perspective?

muhammad-ammar commented on cms/templates/course_outline.html:
```diff
@@ -50,6 +50,40 @@ <h2 class="title title-3">${_("This course was created as a re-run. Some manual
+
+       <p>
+        ${_("You also need to remove peergrading and combinedopenended from the \"Advanced Module List\" in ")}
+        <a href="http://studio.tnl2303.m.sandbox.edx.org/settings/advanced/edX/DemoX/Demo_Course">Studio Advanced Settings</a>.
```
@srpearce There are three pieces of text.
* The heading: this will be visible whenever this box is visible.
* The message with the list of ora1 components: this will be visible if the course has any ora1 components.
* The message with the link to the advance settings page: this will be visible if either `peergrading` or `combinedopenended` are the in the advance modules list.

Can you please suggest the appropriate wording. You can look at how these appear on http://studio.tnl2303.m.sandbox.edx.org/course/edX/DemoX/Demo_Course.

srpearce replied:
OK, I've included some suggested text below.

(Quick question: If users remove "peergrading" and "combinedopenended" before they replace their ORA1 assignments, can they still see the existing ORA1 assignments? I was assuming they can't, so I added "After you replace your ORA1 assignments" to the third text string. If they can still see existing ORA1 assignments, we can just say "EdX strongly recommends that you remove ORA1..." without the "After you replace..." part.)

This course contains ORA1 assignments. EdX no longer supports ORA1. 

Learners cannot see ORA1 assignments. If you want to use ORA in your course, you must replace the following ORA1 assignments with ORA2 assignments. 

(list of assignments)

After you replace your ORA1 assignments, edX strongly recommends that you remove ORA1 from the course advanced settings. Go to the Advanced Settings page, locate the "Advanced Module List” setting, and then delete “peergrading” and “combinedopenended” from the list.

marcotuts commented:
My feedback here would be that the fa-bullhorn icon should change to fa-warning for the error case, but otherwise this looks good to me!

marcotuts commented on cms/static/sass/views/_outline.scss:
```diff
@@ -754,3 +754,8 @@
 
 }
+
+.ora1-components-list {
```
we should add this styling to what exists already in _system-feedback.scss, or use a class from that file that is appropriate. If this is special cased we can add a // CASE: deprecation warning styling element. I would also use a class name that labels this as a deprecation warning instead of stipulating ora1 specifically.

explorerleslie commented:
@muhammad-ammar looking good. Couple of points: 

- I stupidly deleted the ORA 1 component on your sandbox. >.< Could you add it back? I'm assuming you added it manually because it can't be added in the UI anymore. 
- @srpearce course teams can still see and delete ORA 1 components in their course after they turn the two settings off, so the order doesn't matter here (per your question above).
- Definitely change the icon per @marcotuts 's suggestion. 

Otherwise let me see it again when the text is updated, thanks!

muzaffaryousaf commented on cms/static/sass/views/_outline.scss:
```diff
@@ -754,3 +754,4 @@
   }
 
 }
+
```
Remove this line.

muhammad-ammar commented:
@explorerleslie @marcotuts @srpearce @muzaffaryousaf Updated the code as per your feedback. Please continue the review. Thanks

muhammad-ammar commented:
@cptvitamin Can you please do an accessibility review?

srpearce commented:
LGTM

muhammad-ammar commented:
Here is a scenario: One of `peergrading` or `combinedopenended` or both modules are added into `Advance Module List` but there are no ORA1 components present in the course.  In that case i think the below message should be different especially the bold text 

Learners cannot see ORA1 assignments. If you want to use ORA in your course, **you must replace the following ORA1 assignments with ORA2 assignments**.

@explorerleslie @srpearce Should we need to consider this scenario?

srpearce commented:
@muhammad-ammar I think that's a good point - that message will be confusing if the course has the two advanced modules but no ORA1 components. And in that case, even the top message ("This course contains ORA1 assignments. Your Platform Name Here no longer supports ORA1.") would be confusing.

I think we should change that top message to something more generic so that it will work whether or not the course has ORA1 components:

**This course has ORA1 enabled. Platform no longer supports ORA1.**

We would leave out the second part ("Learners cannot see ORA1 assignments...") and just keep the third part ("To avoid errors..."). 

So the messages would be these. (I'm assuming that it's not possible for the course to have ORA1 components but not have the modules.)

Course has ORA1 modules but no components
------------------------------------------------------------------

**This course has ORA1 enabled. Platform no longer supports ORA1.**

To avoid errors, Your Platform Name Here strongly recommends that you remove ORA1 from the course advanced settings. To do this, go to the Advanced Settings page, locate the "Advanced Module List" setting, and then delete "peergrading" and "combinedopenended" from the list.

Course has ORA1 modules and components
------------------------------------------------------------------

**This course has ORA1 enabled. Platform no longer supports ORA1.**

Learners cannot see ORA1 assignments. If you want to use ORA in your course, you must replace the following ORA1 assignments with ORA2 assignments.

(assignment list)

To avoid errors, Your Platform Name Here strongly recommends that you remove ORA1 from the course advanced settings. To do this, go to the Advanced Settings page, locate the "Advanced Module List" setting, and then delete "peergrading" and "combinedopenended" from the list.

muhammad-ammar commented:
> (I'm assuming that it's not possible for the course to have ORA1 components but not have the modules.)

@srpearce It is possible. One can add ORA1 modules then add ORA1 components and then delete ORA1 modules. I am not sure if we really need to think about this case.

muhammad-ammar commented:
@cptvitamin Can you please review the changes for accessibility?

muzaffaryousaf commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -467,6 +470,54 @@ def _get_rerun_link_for_item(course_key):
+
+    Arguments:
+        course_id: course id
+        advanced_modules: advance module list
```
It would be nice if we can update these args just like Returns statement.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -467,6 +470,54 @@ def _get_rerun_link_for_item(course_key):
+            modified=modified_timestamp[0]['modified']
+        )
+    else:
+        return {}
```
Can we make default dict like this ->
```
data = {
'ora1_enabled': False,
'ora1_components': [],
'advance_settings_url': ''
}
```
and return data instead of `{}`.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -340,6 +343,38 @@ def _assert_settings_link_present(response):
+        {'ora1_advance_modules': True, 'ora1_components': True},
+    )
+    @ddt.unpack
+    def test_ora1_deprecated_warning(self, ora1_advance_modules, ora1_components):
```
Maybe `test_verify_ora1_deprecated_warning_message` ?

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,127 @@ def _get_items(self):
+            self.assertEqual(self.course_outline_page.ora1_components_list_heading, self.COMPONENT_LIST_HEADING)
+            self.assertEqual(self.course_outline_page.ora1_components_display_names, components_display_name_list)
+
+    def test_no_warning(self):
```
Maybe `test_no_warning_message_present` ?

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,127 @@ def _get_items(self):
+        self.course_outline_page.visit()
+        self.assertFalse(self.course_outline_page.ora1_deprecated_warning_visible)
+
+    def test_warning(self):
```
Maybe `test_deprecation_message_present` ?

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,127 @@ def _get_items(self):
+        )
+        self.course_fixture.create_xblock(parent_vertical.locator, XBlockFixtureDesc('peergrading', 'Peer'))
+
+    def _verify_warning_info(self, remove_text_state, components_state, components_display_name_list=None):
```
Maybe `_verify_deprecation_warning_info` ?

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,127 @@ def _get_items(self):
+        When I goto course outline
+        Then I see ORA1 deprecated warning
+        And I see correct ORA1 deprecated warning heading text
+        And I don't ORA1 deprecated warning advance modules remove text
```
`don't see ORA1`

muhammad-ammar commented:
@muzaffaryousaf Updated the code as per your feedback.

muhammad-ammar commented:
test this please

marcotuts commented:
:+1:  from my perspective

muhammad-ammar commented:
@symbolist Can you please review the changes?

explorerleslie commented:
@muhammad-ammar please see this course: http://studio.tnl2303.m.sandbox.edx.org/course/course-v1:LeslieX+DemoX+2015_T2. Click the first two links present in the ORA warning. This leads to a unit that doesn't have ORA 1 components. Why is this?

explorerleslie commented:
@muhammad-ammar in the demo course, I just noticed another Advanced Module associated with ORA 1: "openassessment". Do you know what this does? Is this actually ORA 1?

symbolist commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,60 @@ def _get_rerun_link_for_item(course_key):
+        'advance_settings_url': ''
+    }
+
+    modified_timestamp = CourseStructure.objects.filter(course_id=course_id).values('modified')
```
I think the point of caching is to avoid hitting the database. How about we clear the cache on a `course_publish` event instead? That way we do not have to worry about the modification date.

muhammad-ammar commented:
>  in the demo course, I just noticed another Advanced Module associated with ORA 1: "openassessment". Do you know what this does? Is this actually ORA 1?

@explorerleslie This is ORA2

cptvitamin commented on cms/templates/course_outline.html:
```diff
@@ -50,6 +51,45 @@ <h2 class="title title-3">${_("This course was created as a re-run. Some manual
+  %>
+  <div class="wrapper wrapper-alert wrapper-alert-error is-shown">
+    <div class="alert announcement">
+      <i class="feedback-symbol fa fa-warning"></i>
```
In order to make this accessible we need to hide the actual icon from screen readers by adding an ```aria-hidden="true"```  on the ```<i>``` element and then adding some accessible text immediately afterwards like ```<span class="sr">${_("Warning")}</span>```

cptvitamin commented:
Besides my small comment, this looks fine.

muhammad-ammar commented:
> please see this course: http://studio.tnl2303.m.sandbox.edx.org/course/course-v1:LeslieX+DemoX+2015_T2. Click the first two links present in the ORA warning. This leads to a unit that doesn't have ORA 1 components. Why is this?

@explorerleslie Actually ORA1 components are present in that unit but they are unpublished. Below is some detail about what is happening:
1. We have created a unit with ORA1 components but didn't published it. 
2. Then we create another unit with ORA1 components.
3. Now visit the course outline you will see no list of ORA1 components in the deprecated warning message.
4. Now publish the the unit created in `step 2`
5. Now visit the course outline and you will see the list of all(published and unpublished) the ORA1 components. 

We are using [Course Structures](https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/content/course_structures) to get the info about ORA1 components present in course, Course Structures update this info whenever `course_publish` event happened. So when we published the unit in `step 4` above, Course Structures updated its existing info with unpublished and published ORA1 components. And when we renders the course outline page we displayed/listed all(published and unpublished) the ORA1 components. This happens because Course Structures store no information about whether a unit is published or not. 

So this is the whole story why you are seeing unpublished ORA1 components.

stephensanchez commented:
:+1: times 100.

explorerleslie commented:
@muhammad-ammar that makes a ton of sense. @srpearce can you make a minor update to the copy to note that the links that are being shown to units are where published ORA 1 components appear? Maybe just "If you want to use ORA in your course, you must replace the following *published* ORA1 assignments with ORA2 assignments.

srpearce commented:
@explorerleslie That change looks good to me (sans italics, of course). @muhammad-ammar, please use the text below. 

If you want to use ORA in your course, you must replace the following published ORA1 assignments with ORA2 assignments.

muhammad-ammar commented:
@symbolist Please review the changes.

muhammad-ammar commented on cms/templates/course_outline.html:
```diff
@@ -50,6 +51,45 @@ <h2 class="title title-3">${_("This course was created as a re-run. Some manual
+  %>
+  <div class="wrapper wrapper-alert wrapper-alert-error is-shown">
+    <div class="alert announcement">
+      <i class="feedback-symbol fa fa-warning" aria-hidden="true"></i><span class="sr">${_("Warning")}</span>
```
@cptvitamin Is this fine?

muhammad-ammar commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+
+    data['ora1_components'] = ora1_components
+
+    return data
```
@ormsbee Can you please review the above changes. We are fetching ORA1 components here from `CourseStructure` and also caching them.

muhammad-ammar commented on cms/djangoapps/contentstore/signals.py:
```diff
@@ -28,3 +29,12 @@ def listen_for_library_update(sender, library_key, **kwargs):  # pylint: disable
+    Receives signal and remove ora1 deprecated components cache
+    """
+    cache_key = 'ora1.components.{course}'.format(course=course_key)
+    cache.delete(cache_key)  # pylint: disable=maybe-no-member
```
@ormsbee Here we remove the ORA1 components for a particular course upon `course_published` event. In this way we don't need to check the `modified` datetime value from `CourseStructure`

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+    ora1_components = cache.get(cache_key)  # pylint: disable=maybe-no-member
+    if not ora1_components:
+        try:
+            course_structure = CourseStructure.objects.get(course_id=course_id)
```
Can you please use the api.py interface instead of accessing the model directly? Anything that can go through the public interface should.

muhammad-ammar replied:
@ormsbee [api.py](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_structure_api/v0/api.py) exists in LMS. We need to move it into `openedx` to be used in CMS. 
Is it ok to move it into `openedx`?

ormsbee replied:
That sounds fine to me.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+        store = modulestore()
+        ora1_components = []
+        for __, block in ordered_blocks.items():
+            if block['block_type'] in ['peergrading', 'combinedopenended']:
```
This kind of filtering seems generally useful to the course structures API -- can you add it at that level?

ormsbee replied:
Actually, @nasthagiri: does this kind of querying ability overlap with work you're already doing on course structures?

nasthagiri replied:
@ormsbee Two things:
1. My work is for a new student-facing course structures API.  I didn't update the current server-facing API.
2. I have added querying parameters by block_type, but that is only for `block_count` (counting blocks of a certain type in a subtree) and `block_json` (returning the json data for a block type).

I completely agree that it is useful for the course structures API to have inherent support for a filtering capability.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+        for __, block in ordered_blocks.items():
+            if block['block_type'] in ['peergrading', 'combinedopenended']:
+                item = store.get_item(UsageKey.from_string(block['parent']))
+                if not getattr(item, 'is_draft', False):
```
Please add a comment explaining this line (why do we care about `is_draft`)?

muhammad-ammar replied:
We only want to show published ora1 components. I will add the comment.

ormsbee replied:
Why make the distinction though? If it's in the course, draft state or not, shouldn't we still let them know it's a bad idea?

muhammad-ammar replied:
@ormsbee Here is the scenario
1. Create a unit with ORA1 components and publish it
2. Visit the course outline and you will see a list of ORA1 components 
3. Now create two more units with ORA1 components, don't publish them
4. Visit the course outline and you will see the same ORA1 components as you seen in `step 2` 
5. Now publish any of unit from step 3
6. Visit the course outline and you will see all(published and unpublished) the ORA1 components IF THERE IS NO CHECK OF `is_draft` 
7. The problem is behavior is different in `step 4` and `step 6` thats why we added the `is_draft` check to always show the published items

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
     return reverse_course_url('course_rerun_handler', course_key)
 
 
+def _ora1_deprecation_info(course_id, advanced_modules):
```
Is there a way to make this a little more generic in the implementation? So that when we deprecate `foldit` or some other type next, it's just a config change and doesn't involve writing a new method?

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+                        [reverse_usage_url('container_handler', block['parent']), block['display_name']]
+                    )
+
+        cache.set(cache_key, ora1_components, 60 * 5)  # pylint: disable=maybe-no-member
```
One of the reasons I'd prefer that the version/publish_date be used as part of the cache key is so that we don't have to worry about stale keys, invalidating the old key when a publish happens, race conditions, or any of that. As our system evolves and gets more distributed, it is possible that content storage will also become eventually consistent.

muhammad-ammar replied:
```     
modified_timestamp = CourseStructure.objects.filter(course_id=course_id).values('modified')
if modified_timestamp.exists():
    cache_key = 'ora1.components.{course}.{modified}'.format(
         course=course_id,
         modified=modified_timestamp[0]['modified']
    )
else:
    return {}
    
ora1_components = cache.get(cache_key)  # pylint: disable=maybe-no-member
```
@ormsbee will this work?

ormsbee replied:
Something like that for making the key, yeah. Though again, I'd like to see the querying and caching done behind the api.py abstraction. If the method in api.py you called was smart enough to accept as a parameter the block_types you were interested in, then it could just return you that data, and the code here wouldn't need to worry about cache invalidation and such -- that should be a concern of course_structures.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +472,56 @@ def _get_rerun_link_for_item(course_key):
+    cache_key = 'ora1.components.{course}'.format(course=course_id)
+
+    ora1_components = cache.get(cache_key)  # pylint: disable=maybe-no-member
+    if not ora1_components:
```
I think you want to do an explicit check `is not None` here. Otherwise, you'll incur the extra work even if it's an empty list (since that is falsy in Python).

ormsbee commented:
@stephensanchez: Is that a real :+1: as in code-reviewed, or a :+1: as in applause at the PR? I'm guessing the latter, but I wanted to make sure nobody on the ticket misunderstood. :-)

muhammad-ammar commented:
test this please

muhammad-ammar commented on openedx/core/djangoapps/content/course_structures/api/v0/api.py:
```diff
@@ -61,14 +63,39 @@ def course_structure(course_key):
+
+    # If we don't have data stored, generate it and return an error.
+    tasks.update_course_structure.delay(unicode(course_key))
+    raise CourseStructureNotAvailableError
```
@ormsbee After your feedback I have added `block_types` parameter to the course_structure api. This api function now filters the blocks to only include those which have one of these types. And caching is being done inside the api as well. 

Does this part look fine now?

ormsbee replied:
The control flow looks a bit odd here, since I would normally expect the exception handling to be around a much smaller block (i.e. just around the model access, not the whole thing). But in terms of the general logic within the function, yeah, that looks fine, ty.

muhammad-ammar replied:
I will try to improve it.

muhammad-ammar commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +474,52 @@ def _get_rerun_link_for_item(course_key):
     return reverse_course_url('course_rerun_handler', course_key)
 
 
+def _deprecation_info(course_module, deprecated_block_types):
```
@ormsbee I have also made this function generic so that in future we can pass it other deprecated block types.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +474,52 @@ def _get_rerun_link_for_item(course_key):
+
+    store = modulestore()
+    components = []
+    for __, block in structure_data['blocks'].items():
```
If we're ignoring the keys anyway, why not use `.values()` instead of `.items()`?

Also, does this give you the ordering you want?

muhammad-ammar replied:
yes

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +474,52 @@ def _get_rerun_link_for_item(course_key):
+    try:
+        structure_data = api.course_structure(course_module.id, block_types=deprecated_block_types)
+    except errors.CourseStructureNotAvailableError:
+        return data
```
Where does `data` come from?

muhammad-ammar replied:
data is initialized as dict in the start of function.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +474,52 @@ def _get_rerun_link_for_item(course_key):
+
+    data['feature_components'] = components
+
+    return data
```
Where is `data` initialized?

ormsbee commented on openedx/core/djangoapps/content/course_structures/api/v0/api.py:
```diff
@@ -61,14 +63,39 @@ def course_structure(course_key):
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
+        if data:
+            return data
```
1. You can set the cache timeout to forever (`None`), since it should never change for that course+timestamp.
2. Please put an explicit check for `if data is not None:` -- it doesn't really matter in this case, but there are cases where an empty dict or list is ok, and people can miss that more easily if they copy-paste this idiom.

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +474,52 @@ def _get_rerun_link_for_item(course_key):
+
+        ora1_enabled (bool): True if "peergrading" and/or "combinedopenended" are in the Advanced Module List else False
+        ora1_components (list): List of ora1 components and their parent's url
+        advance_settings_url (str): URL to advance settings page
```
Do these comments still apply?

muhammad-ammar replied:
no. i will update the docstring

ormsbee commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -496,6 +547,8 @@ def course_index(request, course_key):
         except (ItemNotFoundError, CourseActionStateItemNotFoundError):
             current_action = None
 
+        ora1_deprecation_info = _deprecation_info(course_module, DEPRECATED_BLOCK_TYPES)
```
Again, is it possible to talk about this as a general deprecation mechanism where the ORA1 types are just listed in configuration, rather than having ora1 specifics in code?

muhammad-ammar replied:
yes..i will update it

ormsbee commented:
Done with this pass. I don't mean to blow up the scope of this, so please let me know if what I'm requesting is difficult to do. Having an ORA1 deprecation is good (goodness knows, I'm looking forward to the funeral), but we don't want to add problem-type specific code every time we deprecate something.

andy-armstrong commented on cms/templates/course_outline.html:
```diff
@@ -50,6 +51,45 @@ <h2 class="title title-3">${_("This course was created as a re-run. Some manual
   </div>
   %endif
+
+  %if ora1_deprecation_info.get('ora1_components') or ora1_deprecation_info.get('feature_enabled'):
```
I'm in agreement with @ormsbee here that the code shouldn't have any special ORA1 knowledge. IMO this deprecation info should be a list of features that are being used but which are deprecated. It would be up to the server to generate a good error message, and then this page would show a list of deprecation alerts.

This is another use case where I would like to see a first class 'feature' extension point that could be registered with the platform. Then a feature could declare itself as being deprecated, and could implement a method to determine if a course is using it, and to then generate an appropriate user message.

muhammad-ammar commented:
@ormsbee Sorry for inconvenience. I made these changes to get your feedback about if CourseStructure querying/filtering and caching is correct. I will surely improve it.

@ormsbee @andy-armstrong  I totally agree with you that there shouldn't be any ora1 specifics in course.py. I will try to remove the ora1 specifics and make the code as much generic as my limited knowledge permits.
Regarding course_outline.html what i have in my mind is to remove the hardcoded ORA1 specific wordings/text from it and pass them from view/config? I am also thinking about moving the `_deprecation_info` function into some other file may be in `helpers.py`
Is it OK? Any suggestions?

ormsbee commented:
Apologies, but I'm not going to have the time to be responsive on this ticket for the next day or two. I defer to @andy-armstrong in terms of review.

muhammad-ammar commented:
@srpearce 

`If you want to use ORA in your course, you must replace the following published ORA1 assignments with ORA2 assignments.
`

We want to update the above message string for different cases as mentioned below

**Case1**: Feature XYZ is deprecated and we must delete its components from course
* {list of XYZ components}

**Case2**: Feature ZYX.1 is deprecated and you must replace its components with ZYX.2 components
* {list of ZYX.1 components}

The goal is to have single message as generic as possible for the above two cases. If possible then can you suggest one?

muhammad-ammar commented:
test this please

muhammad-ammar commented:
@explorerleslie @srpearce As we want to make the warning message generic so that it should work when we deprecated any other module. Can you suggest warning message(s) that would work in various scenarios?

Here we are replacing ORA1 with ORA2, is this the case every time?

srpearce commented:
@muhammad-ammar @explorerleslie Does this work?

Open response assessments version 1 (ORA1) is no longer supported. You must delete the following ORA1 assignments or replace them with ORA2 assignments.

srpearce commented:
Oh, I get it. Let's try this:

This course uses a component that is no longer supported. You must delete or replace the following components.

(Does "component" work, or do we need something more generic than that? I hesitate to use "module" because we don't use that term in other areas of Studio, so users won't know what it is.)

muhammad-ammar commented:
@srpearce The goal is to have generic text messages for warning banner.
We have two scenarios in mind

1. We have deprecated `FeatureX` and it must be replaced with `FeatureZ`
 * List of `FeatureX` components to be replaced with `FeatureZ` components

2. We have deprecated `FeatureA` and there is no replacement
  * List of `FeatureA` components to be deleted

@explorerleslie FYI

muzaffaryousaf commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +471,49 @@ def _get_rerun_link_for_item(course_key):
+        deprecated_blocks (list): list of deprecated blocks info
+    Returns:
+        Dict with following keys:
+        feature_enabled (bool): True if any or all `deprecated_block_types` present in Advanced Module List else False
```
These names does not match with the actual returned dict keys.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -472,6 +474,47 @@ def _get_rerun_link_for_item(course_key):
+    """
+    data = {
+        'block_types': deprecated_block_types,
+        'block_types_enabled': any(component in course_module.advanced_modules for component in deprecated_block_types),
```
I'd rename `component` to `block_type` for consistency.

muzaffaryousaf commented on cms/templates/course_outline.html:
```diff
@@ -50,6 +51,54 @@ <h2 class="title title-3">${_("This course was created as a re-run. Some manual
+                      platform_name=platform_name,
+                      link_start='<a href="{advance_settings_url}">'.format(advance_settings_url=deprecated_blocks_info['advance_settings_url']), link_end="</a>"
+                    )}
+                    </p>
```
Indentation is bit off for `<p>

muzaffaryousaf commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -472,6 +474,47 @@ def _get_rerun_link_for_item(course_key):
+    """
+    Returns deprecation information about `deprecated_block_types`
+    Arguments:
+        course_module (CourseKey): course module
```
I believe this is no more CourseKey, if so, please update the comment.

muzaffaryousaf commented on openedx/core/djangoapps/content/course_structures/api/v0/api.py:
```diff
@@ -23,52 +24,75 @@ def _retrieve_course(course_key):
     Returns:
         The serialized output of the course structure:
             * root: The ID of the root node of the course structure.
-
```
Can we add back these extra lines for more readability.

muzaffaryousaf commented on openedx/core/djangoapps/content/course_structures/models.py:
```diff
@@ -55,4 +55,4 @@ def _traverse_tree(self, block, unordered_structure, ordered_blocks, parent=None
 # Signals must be imported in a file that is automatically loaded at app startup (e.g. models.py). We import them
 # at the end of this file to avoid circular dependencies.
-import signals  # pylint: disable=unused-import
+# import signals  # pylint: disable=unused-import
```
Remove this commented import if you decide to move this to startup.py

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,132 @@ def _get_items(self):
+                                   'page, locate the "Advanced Module List" setting, and then delete the following '
+                                   'modules from the list.')
+
+    def _add_deprecated_advance_modules(self):
```
Can we pass list of modules `['peergrading', 'combinedopenended']` to make it more generic so we can use same method in future.

muzaffaryousaf commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,132 @@ def _get_items(self):
+        self.course_outline_page.visit()
+        self.assertFalse(self.course_outline_page.deprecated_warning_visible)
+
+    def test_deprectaion_warning_message_present(self):
```
`deprecation`.

muzaffaryousaf commented on openedx/core/djangoapps/content/course_structures/api/v0/api.py:
```diff
@@ -23,52 +24,75 @@ def _retrieve_course(course_key):
+        cache_key = 'openedx.content.course_structures.api.v0.api.course_structure.{}.{}.{}'.format(
+            course_key, modified_timestamp[0]['modified'], '_'.join(block_types or [])
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
```
can we add test around the caching to verify that we do not hit database once we set cache.

muhammad-ammar commented:
@muzaffaryousaf Updated the code as per your feedback.

muhammad-ammar commented:
@andy-armstrong Can you please give a second review? Thanks.

muhammad-ammar commented:
@explorerleslie Sandbox is updated.

explorerleslie commented:
:+1:

andy-armstrong commented on lms/djangoapps/course_structure_api/v0/serializers.py:
```diff
@@ -40,37 +40,3 @@ def get_uri(self, course):
         return course_image_url(course)
-
-
-class GradingPolicySerializer(serializers.Serializer):
```
Is this intended to be in a different PR? I wouldn't expect to see this kind of change here.

andy-armstrong commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -1578,3 +1583,139 @@ def _get_items(self):
+
+    def test_no_deprecation_warning_message_present(self):
+        """
+        Scenario: Verify that deprecation warning message is not shown if ORA1 advance modules are not present and
```
Nit: docstrings are wrapped at 80 characters, and wrapped lines are only indented four spaces.

andy-armstrong commented:
@muhammad-ammar This looks fantastic. I'm really pleased with the new generic support for deprecation. :+1:

I am confused why there are changes to some of the course structure API in here. Was that a mistake? It should be moved to a different PR unless there are ties that I'm not seeing with the deprecation logic.

muhammad-ammar commented:
@andy-armstrong Thanks a lot for the review.
Regarding course structure API refactoring, these were necessary because we want to use the `api.py` in CMS. `api.py` was earlier present in LMS. So when we moved `api.py` in `openedex`, `errors.py` and some part of `serializers.py` was also moved.

muhammad-ammar commented:
@andy-armstrong Are we good to merge this?

andy-armstrong commented:
:+1: thanks for the explanation about the API changes. They all look good to me.

muhammad-ammar commented:
@andy-armstrong Thanks a lot

muzaffaryousaf commented:
@muhammad-ammar Thanks for addressing the feedback. This is ready to :ship:

muzaffaryousaf commented:
Here are the stats after performance testing of the new method https://github.com/edx/edx-platform/pull/8450/files#diff-968532c6bba75e1f61d1e1ddb51d7842R477 
Note: The time shows the original time taken by this new method.


Course Title | Course ID | Total Blocks | ORA Blocks | Run 1| Run 2| Run 3| Average
------------ | ------------ | ------------ | ------------ | ------------ | ------------ | ------------- | -------------
Genomic Medicine Gets Personal | GeorgetownX/MEDX202-01/2014_SOND | 1606 | 32 | 0.445359 | 0.195447 | 0.198509 | 0.279771666667
Evaluating Social Programs | MITx/JPAL101_1x/3T2014 | 742 | 22 | 0.215538 | 0.144938 | 0.142242 | 0.167572666667
Age of Globalization | UTAustinX/UT.3.02x/3T2014 | 1259 | 11 | 0.177266 | 0.064799 | 0.065494 | 0.102519666667
Circuits and Electronics | MITx/6.002_4x/3T2014 | 3539 | 4 | 0.255011 | 0.040458 | 0.038747 | 0.111405333333
Introduction to Philosophy: God, Knowledge and Consciousness | MITx/24.00x/2013_SOND | 1000 | 2 | 0.360472 | 0.023832 | 0.022502 | 0.135602
Introduction to Philosophy: God, Knowledge and Consciousness | MITx/24.00_1x/3T2014 | 1015 | 2 | 0.092449 | 0.02285 | 0.024109 | 0.0464693333333
Human Health and Global Environmental Change | HarvardX/PH278x/2013_Spring | 612 | 1 | 0.059008 | 0.027214 | 0.028374 | 0.0381986666667
Greatest Unsolved Mysteries of the Universe | ANUx/ANU-ASTRO1x/1T2014 | 814 | 0 | 0.075921  | 0.013537 | 0.012775 | 0.0340776666667
The Science of Happiness | course-v1:BerkeleyX+GG101x-2+1T2015 | 1292 | 0 | 0.289695 | 0.108428 | 0.251625 | 0.216582666667
Introduction to Linux | LinuxFoundationX/LFS101x/2T2014 | 1862 | 0 | 0.106152 | 0.02137 | 0.017005 | 0.0481756666667
Introduction to Linux | course-v1:LinuxFoundationX+LFS101x.2+1T2015 | 1863 | 0 | 0.274536 | 0.150316 | 0.324813 | 0.249888333333
Mechanics ReView | MITx/8.MReVx/2T2014 | 3678 | 0 | 0.595483 | 0.02546 | 0.02531 | 0.215417666667
Building Mobile Experiences | MITx/21W.789x/1T2014 | 220 | 0 | 0.170285 | 0.037965 | 0.037166 | 0.0818053333333

ormsbee commented:
Do you know where that time is being spent? Mongo queries? CPU?

explorerleslie commented:
@muzaffaryousaf how do these compare with performance from before this new method was introduced?

muzaffaryousaf commented:
@ormsbee  These are new stats with more details, the **value1** in (value1 + value2 ) is time(seconds) spent on the `course_structure_api`, this will vary depending on the total No. of blocks in particular course. The **value2** is time(seconds) spent on mongo queries and this will depend on the No. of ORA blocks in a course. 


Course Title | Course ID | Total Blocks | ORA Blocks | Run 1(course_api + mongo)  | Run 2(course_api + mongo) | Run 3(course_api + mongo) | Average 
------------|------------|------------|------------|------------|------------|------------|------------
Genomic Medicine Gets Personal | GeorgetownX/MEDX202-01/2014_SOND | 1606 | 32 | 0.382389 (0.175144+0.20464) | 0.201347(0.020109+0.18033) | 0.202079(0.020188+0.180907) | 0.261938333333
Evaluating Social Programs | MITx/JPAL101_1x/3T2014 | 742 | 22 | 0.22485(0.092338+0.131173) | 0.150119(0.014146+0.134998) | 0.1454(0.013629+0.130816) | 0.173456333333 
Age of Globalization | UTAustinX/UT.3.02x/3T2014 | 1259 | 11 | 0.155353(0.119243+0.035194) | 0.058745(0.022363+0.035848) | 0.053727(0.019429+0.033737) | 0.089275 
Circuits and Electronics | MITx/6.002_4x/3T2014 | 3539 | 4 | 0.257469(0.232551+0.023917) | 0.042692(0.029258+0.012922) | 0.035555(0.022497+0.012543) | 0.111905333333 
Introduction to Philosophy: God, Knowledge and Consciousness | MITx/24.00x/2013_SOND | 1000 | 2 | 0.293969(0.281345+0.011476) | 0.025569(0.01351+0.011284) | 0.023818(0.013364+0.009681) | 0.114452
Human Health and Global Environmental Change | HarvardX/PH278x/2013_Spring | 612 | 1 | 0.049241(0.044642+0.003916) | 0.024014(0.019847+0.003661) | 0.022212(0.018242+0.003451) | 0.0318223333333 
Greatest Unsolved Mysteries of the Universe | ANUx/ANU-ASTRO1x/1T2014 | 814 | 0 | 0.075921  | 0.013537 | 0.012775 | 0.0340776666667
The Science of Happiness | course-v1:BerkeleyX+GG101x-2+1T2015 | 1292 | 0 | 0.289695 | 0.108428 | 0.251625 | 0.216582666667
Introduction to Linux | LinuxFoundationX/LFS101x/2T2014 | 1862 | 0 | 0.106152 | 0.02137 | 0.017005 | 0.0481756666667
Introduction to Linux | course-v1:LinuxFoundationX+LFS101x.2+1T2015 | 1863 | 0 | 0.274536 | 0.150316 | 0.324813 | 0.249888333333
Mechanics ReView | MITx/8.MReVx/2T2014 | 3678 | 0 | 0.595483 | 0.02546 | 0.02531 | 0.215417666667
Building Mobile Experiences | MITx/21W.789x/1T2014 | 220 | 0 | 0.170285 | 0.037965 | 0.037166 | 0.0818053333333

muzaffaryousaf commented:
@explorerleslie We have measured the performance of newly added method only.

benpatterson commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -340,6 +343,80 @@ def _assert_settings_link_present(response):
+            publish=publish
+        )
+
+    def test_verify_deprecated_warning_message_with_multiple_features(self):
```
Unless it's covered elsewhere, you need a negative test (i.e. that the warning message would NOT appear when there are no deprecated features on the course).

benpatterson replied:
That's a must...whether here or elsewhere. ^

muhammad-ammar replied:
We are testing this in bok-choy tests.

benpatterson commented on lms/djangoapps/course_structure_api/v0/tests.py:
```diff
@@ -336,6 +335,7 @@ def add_block(xblock):
             blocks[unicode(xblock.location)] = {
                 u'id': unicode(xblock.location),
                 u'type': xblock.category,
+                u'parent': None,
```
What's the effect of this change? If there is a new item in the block, I would expect there to be a test with `None` and not `None` ?

muhammad-ammar replied:
In these tests parent will always be set to `None` by api and we verify that the value of parent is indeed set to None.

benpatterson replied:
Yes, I see that the parent will be set to `None` in all the tests. However, is there a circumstance in which it will not be set to `None`? Should you test for that scenario?

muhammad-ammar replied:
it will not set `None` when we request data for certain `block_types` and we are testing those cases in `tests_api.py` where `parent` is set a proper `usage_id`

benpatterson replied:
Ah, gotcha. Thanks @muhammad-ammar

benpatterson commented on openedx/core/djangoapps/content/course_structures/api/v0/tests_api.py:
```diff
@@ -0,0 +1,136 @@
+        """
+        Test setup
+        """
+        # For some reason, `listen_for_course_publish` is not called when we run
```
Is this an issue with mock being set elsewhere?

muhammad-ammar replied:
I think yes.

benpatterson commented on openedx/core/djangoapps/content/course_structures/api/v0/tests_api.py:
```diff
@@ -0,0 +1,136 @@
+        self.video = ItemFactory.create(
+            parent_location=self.vertical.location, category="html", display_name="My HTML"
+        )
+
```
It's a better pattern to use self.addCleanup for the teardown tasks.

muhammad-ammar replied:
agreed

benpatterson commented on openedx/core/djangoapps/content/course_structures/api/v0/tests_api.py:
```diff
@@ -0,0 +1,136 @@
+
+    def test_course_structure_with_block_types(self):
+        """
+        Verify that course_structure returns info for required block_types only when specific block_types are requested.
```
Must: I think you need additional tests here. For example, have a document with blocks like 'html', and 'video', but only request 'video'. Then assert only 'video' is returned. Another case, have a document wiht certain blocks ('foo') but request blocks that simply are not in the document ('bar'), ensure the return is what you expect (empty?).

muhammad-ammar replied:
We are already testing the first case, as you can see in `CourseStructureApiTests.setUp` we have created `chapter`, `sequential`, `vertical`, `video` and `html` block types and then in test we are requesting only two(`video` and `html`) block types. 

I will add the test for second case.

muhammad-ammar commented:
@benpatterson I have addressed all of your comments. Thanks for the feedback.

benpatterson commented:
I called out in the JIRA ticket who else I'd want to hear from, but  :+1: from my perspective.

ormsbee commented:
Right now, the time taken increases linearly with the number of deprecated items found. Can we skip the modulestore instantiation for that and skip the `has_published_version()` check? I realize that changes behavior somewhat, but it still seems acceptable to me... @explorerleslie?

I'm guessing that you're seeing such high numbers for even the cached version of the large split courses because of the overhead of fetching the course structures. If we can eliminate the need for the modulestore published version check, we could drop that overhead altogether. If we need to keep it, would a `bulk_operation()` + `get_items()` help reduce the modulestore impact?

muhammad-ammar commented:
test this please

muzaffaryousaf commented:
test this please

explorerleslie commented:
@ormsbee thanks. we talked about this at standup and I'm fine with your proposed change.

muhammad-ammar commented:
@ormsbee We have removed the `has_published_version()` check, now we will show published and unpublished deprecated blocks. This means now there will be no queries to mongo.

muhammad-ammar commented:
@muzaffaryousaf Can you please review the latest changes.

muzaffaryousaf commented on openedx/core/djangoapps/content/course_structures/api/v0/api.py:
```diff
@@ -60,15 +67,41 @@ def course_structure(course_key):
+                structure['blocks'] = required_blocks
+
+            data = CourseStructureSerializer(structure).data
+            cache.set(cache_key, data, None)  # pylint: disable=maybe-no-member
```
@e0d We have added a new cache item here, the data for this item will be the course structure tree and we will be updating it on every course publish event. @benpatterson told me to show this to devOps, does it looks okay ? 
Thanks

muzaffaryousaf replied:
@e0d A few more info about the cache item is here:

1.  A new item will be created every time this api is called after course publish event.
2.  The item will consist of a list of dicts about all or some of the blocks in the course.
3.  There is no expiry for the item that we put in the cache.

Are there any scaling concerns around this ?

muzaffaryousaf replied:
@fredsmith Would you able to take a look on comments ^^ .

fredsmith replied:
The cache should be able to handle the fact that there's no expiration - least recently used items will automatically expire out of the cache. 

As this is a studio page, I don't see an issue with load related to creating the records. 

:+1:

ormsbee commented:
:+1:
